README.md
# Daily Weather Graph

This is a script (and associated CSS styles) for creating an SVG graph of daily weather observations.
Plots daily rainfall totals, low and high temperatures, and daily high wind gusts and their directions. Written with the D3.js library version 4.

![Example image](example.png)

You create a new graph like this, where `config` is a configuration object:
```javascript
var graph = new DailyWeatherGraph(config);
```

The current options you **must** specify are:

+ ***container*** - the HTML element where the chart will be drawn within.
+ ***data*** - the data for the chart - see <a href="demo.js">demo.js</a> for an example of what format this should be in.

Additional options you can specify are:

+ ***width*** - width of the chart in pixels, default `800`.
+ ***height*** - height of the chart in pixels, default `375`.
+ ***dateFormat*** - date format string for parsing the dates in your dataset. See <a href="https://github.com/d3/d3/blob/master/API.md#time-formats-d3-time-format">d3.timeFormat</a> for accepted formats. Default `'%y-%m-%d'`.
+ ***missingValue*** - the value used in the data to represent a missing value, default `null`.
+ ***missingValueText*** - the label to use in the chart for missing values. Defaults to `'n/a'`.
+ ***temperatureUnit*** - temperature unit. Either `'C'` for Celsius or `'F'` for Fahrenheit. Defaults to `'C'`.
+ ***rainfallUnit*** - rainfall unit. Either `'mm'` or `'inches'`. Defaults to `'mm'`.
+ ***windUnit*** - wind gust unit. One of `'km/h'`, `'mph'`, `'knots'`, `'m/s'`. Defaults to `'km/h'`.


CHANGELOG.md
# 2.0.4

- Fix bug where using `inches` for `rainfallUnit` would not set up the scale correctly.

# 2.0.3

- Remove remaining hardcoded styles from js, create css classes instead
- Fix cursor positioning bug
- Make sure temperature values are rounded before being used

# 2.0.2

- Fix code bug

# 2.0.1

- Fix incorrect typings

# 2.0.0

- Make data object keys lowerCamelCase, not UpperCamelCase
- Add typings file for use with typescript
- Change default `missingValue` config option to `null` from `-100`
- Expect `container` config option to be plain HTMLElement, not a D3 selection

# 1.0.0

- Add npm package
- Code tidy up
- Upgrade to D3 version 4
- CSS classes now have `dwg-` prefix

test.js
var data = [
    {Date: "2015-06-12", HighTemperature: 17, LowTemperature: 10, Rainfall: 1.4, HighWindGust:52, HighWindGustBearing: 236},
    {Date: "2015-06-13", HighTemperature: 16, LowTemperature: 13, Rainfall: 0, HighWindGust: 41,HighWindGustBearing: 240},
    {Date: "2015-06-14", HighTemperature: 16, LowTemperature: 12, Rainfall: 2.4, HighWindGust: 41,HighWindGustBearing: 233},
    {Date: "2015-06-15", HighTemperature: 16, LowTemperature: 10, Rainfall: 8.4, HighWindGust: 40,HighWindGustBearing: 261},
    {Date: "2015-06-16", HighTemperature: 12, LowTemperature: 4, Rainfall: 1, HighWindGust: 36,HighWindGustBearing: 229},
    {Date: "2015-06-17", HighTemperature: 14, LowTemperature: 4, Rainfall: 0, HighWindGust: 10,HighWindGustBearing: 74},
    {Date: "2015-06-18", HighTemperature: 15, LowTemperature: 11, Rainfall: 0, HighWindGust: 24,HighWindGustBearing: 321},
    {Date: "2015-06-19", HighTemperature: 18, LowTemperature: 12, Rainfall: 0, HighWindGust: 35, HighWindGustBearing: 340},
    {Date: "2015-06-20", HighTemperature: 17, LowTemperature: 15, Rainfall: 9.6, HighWindGust: 36, HighWindGustBearing: 302},
    {Date: "2015-06-21", HighTemperature: 15, LowTemperature: 9, Rainfall: 0.2, HighWindGust: 33, HighWindGustBearing: 238},
    {Date: "2015-06-22", HighTemperature: 14, LowTemperature: 4, Rainfall: 0.4, HighWindGust: 31, HighWindGustBearing: 218},
    {Date: "2015-06-23", HighTemperature: 11, LowTemperature: 1, Rainfall: 0, HighWindGust: 26, HighWindGustBearing: 222},
    {Date: "2015-06-24", HighTemperature: 11, LowTemperature: 0, Rainfall: 0, HighWindGust: 31, HighWindGustBearing: 213},
    {Date: "2015-06-25", HighTemperature: 13, LowTemperature: 4, Rainfall: 0, HighWindGust: 43, HighWindGustBearing: 241},
    {Date: "2015-06-26", HighTemperature: 16, LowTemperature: 9, Rainfall: 0.4, HighWindGust: 40, HighWindGustBearing: 230},
    {Date: "2015-06-27", HighTemperature: 15, LowTemperature: 10, Rainfall: 2.4, HighWindGust: 43, HighWindGustBearing: 250},
    {Date: "2015-06-28", HighTemperature: 16, LowTemperature: 9, Rainfall: 0.6, HighWindGust: 31, HighWindGustBearing: 240},
    {Date: "2015-06-29", HighTemperature: 16, LowTemperature: 10, Rainfall: 1.2, HighWindGust: 45, HighWindGustBearing: 250},
    {Date: "2015-06-30", HighTemperature: 16, LowTemperature: 11, Rainfall: 1.8, HighWindGust: 38, HighWindGustBearing: 228},
    {Date: "2015-07-01", HighTemperature: 16, LowTemperature: 9, Rainfall: 0.8, HighWindGust: 47, HighWindGustBearing: 250},
    {Date: "2015-07-02", HighTemperature: 14, LowTemperature: 7, Rainfall: 0.2, HighWindGust: 12, HighWindGustBearing :47},
    {Date: "2015-07-03", HighTemperature: 16, LowTemperature: 7, Rainfall: 1.2, HighWindGust: 29, HighWindGustBearing: 65},
    {Date: "2015-07-04", HighTemperature: 16, LowTemperature: 12, Rainfall: 5.6, HighWindGust: 22, HighWindGustBearing: 107},
    {Date: "2015-07-05", HighTemperature: 15, LowTemperature: 11, Rainfall: 0.2, HighWindGust: 47, HighWindGustBearing: 248},
    {Date: "2015-07-06", HighTemperature: 16, LowTemperature: 10, Rainfall: 1.8, HighWindGust: 28, HighWindGustBearing: 312},
    {Date: "2015-07-07", HighTemperature: 17, LowTemperature: 11, Rainfall: 41.4, HighWindGust: 33, HighWindGustBearing: 330},
    {Date: "2015-07-08", HighTemperature: 12, LowTemperature: 3, Rainfall: 0.8, HighWindGust: 31, HighWindGustBearing: 214},
    {Date: "2015-07-09", HighTemperature: 10, LowTemperature: 3, Rainfall: 0, HighWindGust: 45, HighWindGustBearing: 225},
    {Date: "2015-07-10", HighTemperature: 10, LowTemperature: 0, Rainfall: 0, HighWindGust: 28, HighWindGustBearing: 215},
    {Date: "2015-07-11", HighTemperature: 12, LowTemperature: 0, Rainfall: 0, HighWindGust: 33, HighWindGustBearing: 150}
];

var graph = new DailyWeatherGraph({

    data: data,
    container: d3.select('#test'),
    width: 900,
    height: 450,
    dateFormat: '%Y-%m-%d',
    missingValue: -100,
    missingValueText: 'n/a',

});


DailyWeatherGraph.js
(function() {

'use strict';

window.DailyWeatherGraph = function DailyWeatherGraph(cfg) {

    var defaults = {
        width: 800,
        height: 375,
        dateFormat: '%Y-%m-%d',
        missingValue: null,
        missingValueText: 'n/a',
        temperatureUnit: 'C',
        rainfallUnit: 'mm',
        windUnit: 'km/h'
    };

    // Clone config object
    var config = _cloneShallow(cfg);

    // Merge defaults with provided config object
    for (var prop in defaults) {
        if (!_hasOwn(config, prop))
            config[prop] = defaults[prop];
    }

    var numPoints = config.data.length;

    // Dimension settings: widths, heights, margins, etc.
    var dim = {
        marginTop: 10,
        marginRight: 10,
        marginBottom: 30,
        marginLeft: 10,

        regionSpacing: 30,
        windRegionHeight: 70,
        rainRegionHeightProp: 0.375,
        tempRegionHeightProp: 0.625,

        windSize: 50,
        windCircleRadius: 12
    };

    // Calculate dimensions of actual plot area
    dim.plotAreaWidth = config.width - dim.marginLeft - dim.marginRight;
    dim.plotAreaHeight = config.height - dim.marginTop - dim.marginBottom;

    var remainingHeight = (config.height - dim.marginTop -
        dim.marginBottom - 2 * dim.regionSpacing -
        dim.windRegionHeight);

    dim.rainRegionHeight = remainingHeight * dim.rainRegionHeightProp;
    dim.tempRegionHeight = remainingHeight * dim.tempRegionHeightProp;

    // Make sure the graph container is empty
    config.container.html('');

    // Insert the SVG element and apply the margins
    var svgEl = config.container.append('svg')
        .attr('width', config.width)
        .attr('height', config.height)
        .attr('class',  _className('root'));

    // Background rect for whole SVG area
    svgEl.append('rect')
        .attr('class',  _className('graph-background'))
        .attr('x', 0)
        .attr('y', 0)
        .attr('width', config.width)
        .attr('height', config.height);

    // Background rect for whole plot area
    var plotArea = svgEl.append('rect')
        .attr('class',  _className('plot-area-background'))
        .attr('x', dim.marginLeft)
        .attr('y', dim.marginTop)
        .attr('width', dim.plotAreaWidth)
        .attr('height', dim.plotAreaHeight)
        .attr('pointer-events', 'all');

    // Key function for binding data to elements
    var keyFunc = function(d) {
        return d.Date;
    };

    // Create definitions for wind icons & arrows
    var defs = svgEl.append('defs');
    _createWindIndicatorDefs(defs, dim);

    // Get the SVG context we will be drawing in (simply applies the margins).
    var svg = svgEl.append('g')
        .attr('class', _className('plot-area'))
        .attr('transform', 'translate(' + dim.marginLeft + ',' + dim.marginTop + ')')
        .attr('pointer-events', 'none');

    // The x scale. Make sure there is good spacing between points.
    var xScale = d3.scaleBand()
        .domain(_mapVariable(config.data, 'Date'))
        .rangeRound([0, dim.plotAreaWidth], 0.25);

    _drawWindRegion(svg, xScale, config, dim, keyFunc);
    _drawTemperatureRegion(svg, xScale, config, dim, keyFunc);
    _drawRainRegion(svg, xScale, config, dim, keyFunc);

    // Insert day separator(s) between months. To do so, need to get the positions
    // of the two points and average them, then add half the interval width.

    var dateFormat = d3.timeFormat(config.dateFormat);
    var dateParser = d3.timeParse(dateFormat);

    var daysOfMonth = config.data.map(function(d) {
        return dateParser(d.Date).getDate();
    });

    var indexOfNewMonth = daysOfMonth.indexOf(1);

    while (indexOfNewMonth !== -1) {

        if (indexOfNewMonth < numPoints && indexOfNewMonth > 0) {

            var prevMonthDate = data[indexOfNewMonth - 1].Date,
                currMonthDate = data[indexOfNewMonth].Date,
                newMonthLineX = (xScale(prevMonthDate) + xScale(currMonthDate)) / 2 + xScale.bandwidth() / 2;

            svg.append('line')
                .attr('class', _className('vertical-separator'))
                .attr('x1', newMonthLineX)
                .attr('x2', newMonthLineX)
                .attr('y1', dim.marginTop)
                .attr('y2', dim.plotAreaHeight + 20);
        }

        indexOfNewMonth = daysOfMonth.indexOf(1, indexOfNewMonth + 1);

    }

    // Draw cursor (vertical guideline when mousing over)
    var cursor = svg.append('line')
        .attr('x1', 0)
        .attr('x2', 0)
        .attr('y1', dim.marginTop)
        .attr('y2', dim.plotAreaHeight)
        .attr('id', 'graph-cursor')
        .style('stroke', '#BBB')
        .style('stroke-width', '1px')
        .style('stroke-opacity', 0);

    // Create the x axis, with no ticks showing.
    var xAxis = d3.axisBottom()
        .scale(xScale)
        .tickSizeOuter(0)
        .tickSizeInner(0)
        .tickFormat(function(date) {
            var dateParts = date.split('-');
            return parseInt(dateParts[2], 10);
        });

    // Draw the x axis and labels
    svg.append('g')
        .attr('class', _className('axis', 'x-axis'))
        .attr('transform', 'translate(0,' + dim.plotAreaHeight + ')')
        .call(xAxis)
        .selectAll('text')
        .attr('dy', '1.4em')
        .style('font-size', '11px');

    plotArea.on('mousemove', function() {

        // On mousemove, move the graph cursor to follow the mouse pointer
        var pos = d3.mouse(svgEl.node());
        var pt = svgEl.node().createSVGPoint();
        pt.x = pos[0];
        pt.y = pos[1];

        var point = pt.matrixTransform(svgEl.node().getScreenCTM().inverse());

        cursor.attr('x1', point.x);
        cursor.attr('x2', point.x);

    });

    plotArea.on('mouseover', function() {
        // On mouseover, show the graph cursor
        cursor.style('stroke-opacity', 0.8);
    });

    plotArea.on('mouseout', function() {
        // On mouseout, hide the graph cursor
        cursor.style('stroke-opacity', 0);
    });

};



function _createWindIndicatorDefs(defs, dim) {

    var windSymbol = defs.append('g')
        .attr('id', 'daily-weather-graph-wind-icon');

    // Use circle shape
    windSymbol.append('circle')
        .attr('cx', dim.windSize / 2)
        .attr('cy', dim.windSize / 2)
        .attr('stroke-width', '1.3px')
        .attr('r', dim.windCircleRadius);

    var indicator_y0 = dim.windSize / 2 - dim.windCircleRadius + 4,
        indicator_length = 8,
        indicator_y1 = indicator_y0 - indicator_length,
        indicator_x0 = dim.windSize / 2 - indicator_length / 2,
        indicator_x1 = dim.windSize / 2 + indicator_length / 2;

    defs.append('g')
        .attr('id', 'daily-weather-graph-wind-indicator')
        .append('path')
        .attr('d', 'M' + indicator_x0 + ',' + indicator_y1 +
            ' L' + indicator_x1 + ',' + indicator_y1 +
            ' L' + dim.windSize / 2 + ',' + indicator_y0 +
            ' L' + indicator_x0 + ',' + indicator_y1);

}



function _drawRainRegion(svg, xScale, config, dim, keyFunc) {

    var lowestRainScaleMax = config.rainUnit === 'inches' ? 0.4 : 10,
        rainDecimals = (config.rainUnit === 'inches' ? 2 : 1);

    // Filter any NA values
    var rainTotals = _mapVariable(config.data, 'Rainfall').filter(function(r) {
        return r !== config.missingValue;
    });

    // The rain scale. Leave space at top for labels.
    var rainScale = d3.scaleLinear()
        .range([0, dim.rainRegionHeight - 30])
        .domain([0, Math.max(lowestRainScaleMax, d3.max(rainTotals))])
        .clamp(true);

    var rainInterpolator = _getRainInterpolator(config),
        rainRegion = svg.append('g').attr('class', _className('rain-region'));

    rainRegion.append('rect')
        .attr('class',  _className('rain-region-background'))
        .attr('x', 0)
        .attr('y', dim.plotAreaHeight - dim.rainRegionHeight - dim.regionSpacing / 2)
        .attr('width', dim.plotAreaWidth)
        .attr('height', dim.rainRegionHeight + dim.regionSpacing / 2);

    var rainBars = rainRegion.append('g').attr('class',  _className('rain-bars')),
        rainLabels = rainRegion.append('g').attr('class',  _className('rain-labels'));

    // Insert the rainfall bars
    rainBars.selectAll('rect.rain-bar')
        .data(config.data, keyFunc)
        .enter()
        .append('rect')
        .attr('class',  _className('rain-bar'))
        .style('fill', function(d) {
            return (d.Rainfall === 0 || d.Rainfall === config.missingValue ?
                'none' :
                rainInterpolator(d.Rainfall));
        })
        .style('stroke', function(d) {
            if (d.Rainfall === 0 || d.Rainfall === config.missingValue)
                return 'none';
            var col = rainInterpolator(d.Rainfall);
            return d3.rgb(col).darker(0.8);
        })
        .attr('x', function(d) { return xScale(d.Date);})
        .attr('y', function(d) {
            return dim.plotAreaHeight - (d.Rainfall === config.missingValue ?
                0 :
                rainScale(d.Rainfall));
        })
        .attr('width', xScale.bandwidth())
        .attr('height', function(d) {
            return (d.Rainfall === config.missingValue ?
                0 :
                rainScale(d.Rainfall));
        });

    // Insert the bar labels
    // For zero rainfall, insert a small dot instead, and increase its font size.
    // Remember to check values are all at least 0.
    rainLabels.selectAll('text.rain-label')
        .data(config.data, keyFunc)
        .enter()
        .append('text')
        .attr('class', function(d) {
            return (d.Rainfall === config.missingValue ?
				_className('missing-label') :
				_className('rain-label'));
        })
        .attr('text-anchor', 'middle')
        .attr('x', function(d) {return xScale(d.Date) + xScale.bandwidth() / 2;})
        .attr('y', function(d) {
            var r = (d.Rainfall === config.missingValue ? 0 : d.Rainfall);
            return dim.plotAreaHeight - rainScale(r) - 8;
        })
        .style('font-size', function(d) {
            return d.Rainfall === 0 ? '15px' : '';
        })
        .text(function(d) {
            if (d.Rainfall > 0) {
                var z = Math.pow(10, rainDecimals);
                var label = Math.round(z * d.Rainfall) / z;
                // For 2 dp values, remove leading zero
                //  so labels are not too crammed together
                if (label < 0.1) {
                    label = ('' + label).substring(1);
                }
                return label;
            }
            return d.Rainfall === 0 ? '.' : config.missingValueText;
        });

    // Draw rainfall label in top left of rain region
    rainRegion.append('text')
        .attr('class',  _className('section-label'))
        .attr('x', 5)
        .attr('y', dim.plotAreaHeight - dim.rainRegionHeight - dim.regionSpacing / 2)
        .attr('dy', '1.3em')
        .text('Daily rainfall (' + config.rainfallUnit + ')');

    // Draw rain total label at top-right of rain region
    rainRegion.append('text')
        .attr('class',  _className('statistic-label'))
        .attr('x', dim.plotAreaWidth - 5)
        .attr('y', dim.plotAreaHeight - dim.rainRegionHeight - dim.regionSpacing / 2)
        .attr('text-anchor', 'end')
        .attr('dy', '1.3em')
        .text('Total: ' + d3.sum(rainTotals).toFixed(rainDecimals));

    // Add line to separate rain region from temperature region
    rainRegion.append('line')
        .attr('class',  _className('region-separator'))
        .attr('transform', 'translate(0,' + (dim.plotAreaHeight - dim.rainRegionHeight - dim.regionSpacing / 2) + ')')
        .attr('x1', 0)
        .attr('x2', dim.plotAreaWidth)
        .attr('y1', 0)
        .attr('y2', 0);

}



function _drawTemperatureRegion(svg, xScale, config, dim, keyFunc) {

    // The temperature scale
    var tempScale = d3.scaleLinear()
        .range([dim.plotAreaHeight - dim.rainRegionHeight - dim.regionSpacing,
                dim.windRegionHeight + dim.regionSpacing]);

    // Calculate the range of the temperature data.
    var highTemps = _mapVariable(config.data, 'HighTemperature'),
        lowTemps = _mapVariable(config.data, 'LowTemperature');

    var validTemps = highTemps.concat(lowTemps).filter(function(t) {
        return t !== config.missingValue;
    });

    var tempScaleLimits = d3.extent(validTemps),
        tempScaleRange = tempScaleLimits[1] - tempScaleLimits[0];

    // Pad out the temperature axis limits, more so at the top.
    // If there is no variation throughout the period, use +/- 1 degree
    tempScaleLimits[0] = Math.floor(tempScaleLimits[0]) - Math.max(1, 0.15 * tempScaleRange);
    tempScaleLimits[1] = Math.ceil(tempScaleLimits[1]) + Math.max(1, 0.25 * tempScaleRange);

    // Set the domain of the temperature scale
    tempScale.domain(tempScaleLimits);

    // D3 line function for calculating temperature path coordinates
    var lowTemperatureLine = d3.line()
        .x(function(d) { return xScale(d.Date) + xScale.bandwidth() / 2; })
        .y(function(d) { return tempScale(d.LowTemperature); });

    var highTemperatureLine = d3.line()
        .x(function(d) { return xScale(d.Date) + xScale.bandwidth() / 2; })
        .y(function(d) { return tempScale(d.HighTemperature); });

    lowTemperatureLine.defined(function(d) {
        return d.LowTemperature !== config.missingValue;
    });

    highTemperatureLine.defined(function(d) {
        return d.HighTemperature !== config.missingValue;
    });

    var tempRegion = svg.append('g').attr('class',  _className('temperature-region'));

    tempRegion.append('rect')
        .attr('class',  _className('temperature-region-background'))
        .attr('x', 0)
        .attr('y', dim.windRegionHeight + dim.regionSpacing / 2)
        .attr('width', dim.plotAreaWidth)
        .attr('height', dim.tempRegionHeight + dim.regionSpacing);

    // Draw the temperature lines
    tempRegion.append('path')
        .attr('class',  _className('temperature-low-path'))
        .attr('d', lowTemperatureLine(config.data));

    tempRegion.append('path')
        .attr('class',  _className('temperature-high-path'))
        .attr('d', highTemperatureLine(config.data));

    var tempHighLabels = tempRegion.append('g').attr('class',  _className('temperature-high-labels')),
        tempLowLabels = tempRegion.append('g').attr('class',  _className('temperature-low-labels'));

    // Draw the high temperature labels
    tempHighLabels.append('g')
        .selectAll('text')
        .data(config.data, keyFunc)
        .enter()
        .append('text')
        .attr('class',  _className('temperature-high-label'))
        .attr('x', function(d) { return xScale(d.Date) + xScale.bandwidth() / 2; })
        .attr('y', function(d) { return tempScale(d.HighTemperature); })
        .attr('dy', '0.35em')
        .attr('text-anchor', 'middle')
        .text(function(d) {
            return (d.HighTemperature === config.missingValue ? '' : d.HighTemperature);
        });

    // Draw the low temperature labels
    tempLowLabels.append('g')
        .selectAll('text')
        .data(config.data, keyFunc)
        .enter()
        .append('text')
        .attr('class',  _className('temperature-low-label'))
        .attr('x', function(d) { return xScale(d.Date) + xScale.bandwidth() / 2; })
        .attr('y', function(d) { return tempScale(d.LowTemperature); })
        .attr('dy', '0.35em')
        .attr('text-anchor', 'middle')
        .text(function(d) {
            return (d.LowTemperature === config.missingValue ? '' : d.LowTemperature);
        });

    // Draw the temperature region and wind region separator
    tempRegion.append('line')
        .attr('class',  _className('region-separator'))
        .attr('transform', 'translate(0,' + (dim.windRegionHeight + dim.regionSpacing / 2) + ')')
        .attr('x1', 0)
        .attr('x2', dim.plotAreaWidth)
        .attr('y1', 0)
        .attr('y2', 0);

    // Draw temperature label in top left of temperature region
    tempRegion.append('text')
        .attr('class',  _className('section-label'))
        .attr('x', 5)
        .attr('y', dim.windRegionHeight + dim.regionSpacing / 2)
        .attr('dy', '1.3em')
        .text('High and low temperatures (°' + config.temperatureUnit + ')');

}



function _drawWindRegion(svg, xScale, config, dim, keyFunc) {

    // Draw the arrows with <use> elements. They need to be rotated according to the
    // wind direction at a particular point.

    var windInterpolator = _getWindInterpolator(config);
    var y = 35;

    var windRegion = svg.append('g').attr('class',  _className('wind-region'));

    windRegion.append('rect')
        .attr('class',  _className('wind-region-background'))
        .attr('x', 0)
        .attr('y', 0)
        .attr('width', dim.plotAreaWidth)
        .attr('height', dim.windRegionHeight + dim.regionSpacing / 2);

    var windIcons = windRegion.append('g').attr('class',  _className('wind-icons')),
        windArrows = windRegion.append('g').attr('class',  _className('wind-arrows')),
        windGustLabels = windRegion.append('g').attr('class',  _className('wind-gust-labels')),
        windDirLabels = windRegion.append('g').attr('class',  _className('wind-direction-labels'));

    windIcons.selectAll('wind-icon')
        .data(config.data, keyFunc)
        .enter()
        .append('use')
        .attr('xlink:href', '#daily-weather-graph-wind-icon')
        .attr('transform', function(d) {
            if (d.HighWindGustBearing === config.missingValue)
                return '';
            var trans_x = (xScale(d.Date) + xScale.bandwidth() / 2);
            return 'translate(' + (trans_x - dim.windSize / 2) + ',' + y + ')';
        })
        .style('fill', function(d) {
            return (d.HighWindGust === config.missingValue ?
                'none' :
                windInterpolator(d.HighWindGust));
        })
        .style('stroke', function(d) {
            if (d.HighWindGust === config.missingValue)
                return 'none';
            var col = windInterpolator(d.HighWindGust);
            return d3.rgb(col).darker(1.25);
        });

    // Draw wind indicators

    windArrows.selectAll('wind-arrow')
        .data(config.data, keyFunc)
        .enter()
        .append('use')
        .attr('xlink:href', '#daily-weather-graph-wind-indicator')
        .attr('transform', function(d) {
            if (d.HighWindGustBearing === config.missingValue)
                return '';
            var trans_x = (xScale(d.Date) + xScale.bandwidth() / 2),
                rotate = 45 * Math.round(d.HighWindGustBearing / 45);
            return 'translate(' + (trans_x - dim.windSize / 2) + ',35) rotate(' +
                rotate + ' ' + dim.windSize / 2 + ' ' + dim.windSize / 2 + ')';
        })
        .style('fill', function(d) {
            return (d.HighWindGustBearing === config.missingValue || Math.round(d.HighWindGust) === 0 ?
                'none' :
                '#000000');
        })
        .style('stroke', function(d) {
            return (d.HighWindGustBearing === config.missingValue || Math.round(d.HighWindGust) === 0 ?
                'none' :
                '#FFFFFF');
        });

    // Draw the labels for wind direction and speed
    windGustLabels.selectAll('text.wind-gust-label')
        .data(config.data, keyFunc)
        .enter()
        .append('text')
        .attr('class', function(d) {
            return (d.HighWindGust === config.missingValue ?
				_className('na-label') :
				_className('wind-gust-label'));
        })
        .attr('x', function(d) { return xScale(d.Date) + xScale.bandwidth() / 2; })
        .attr('y', y + dim.windSize / 2)
        .attr('dy', '0.35em')
        .attr('text-anchor', 'middle')
        .text(function(d) {
            return (d.HighWindGust === config.missingValue ?
                config.missingValueText :
                Math.round(d.HighWindGust));
        });

    windDirLabels.selectAll('text.wind-direction-label')
        .data(config.data, keyFunc)
        .enter()
        .append('text')
        .attr('class',  _className('wind-direction-label'))
        .attr('x', function(d) { return xScale(d.Date) + xScale.bandwidth() / 2; })
        .attr('y', y)
        .attr('text-anchor', 'middle')
        .text(function(d) {
            return (d.HighWindGustBearing === config.missingValue || Math.round(d.HighWindGust) === 0 ?
                '' :
                _getWindDirection(d.HighWindGustBearing));
        });

    // Draw wind label in top left of wind region
    windRegion.append('text')
        .attr('class',  _className('section-label'))
        .attr('x', 5)
        .attr('y', 5)
        .attr('dy', '0.8em')
        .text('Peak wind gust (' + config.windUnit + ')');

}



/*---------------------------- Utility functions ----------------------------*/

function _className(className) {
    return 'dwg-' + className;
}

function _hasOwn(obj, prop) {
    return Object.prototype.hasOwnProperty.call(obj, prop);
}

function _cloneShallow(obj) {

    var clone = {};

    for (var prop in obj) {
        if (_hasOwn(obj, prop))
            clone[prop] = obj[prop];
    }

    return clone;

}

function _mapVariable(objects, prop) {
    return objects.map(function(obj) {
        return obj[prop];
    });
}

function _getWindDirection(bearing) {
    var directions = ['N', 'NE', 'E', 'SE', 'S', 'SW', 'W', 'NW', 'N'];
    return directions[Math.round(bearing / 45)];
}

function _getRainInterpolator(config) {

    var maxAt = config.rainUnit === 'inches' ? 2 : 50,
        interpolator = d3.interpolateRgb('#4BA7EB', '#0023AD');

    return function(value) {
        var x = Math.max(0, Math.min(value / maxAt, 1));
        return interpolator(x);
    };

}

function _getWindInterpolator(config) {

    var multiplier,
        strongAt = 50,
        maxAt = 100;

    switch (config.windUnit) {
        case 'knots':
           multiplier = 1 / 0.54;
           break;
        case 'mph':
           multiplier = 1 / 0.62137;
           break;
        case 'm/s':
           multiplier = 3.6;
           break;
        default:
            multiplier = 1;
    }

    var interpolators = {
        light: d3.interpolateRgb('#222222', '#808080'),
        strong: d3.interpolateRgb('#C76100', '#C70000')
    };

    return function(value) {

        var kmh = value * multiplier;
        var interpolatorName, interpolatorMin, interpolatorMax;

        if (kmh < strongAt) {
            interpolatorName = 'light';
            interpolatorMin = 0;
            interpolatorMax = strongAt;
        } else {
            interpolatorName = 'strong';
            interpolatorMin = strongAt;
            interpolatorMax = maxAt;
        }

        var x = (value - interpolatorMin) / (interpolatorMax - interpolatorMin);
        x = Math.max(0, Math.min(x, 1));
        return interpolators[interpolatorName](x);

    };

}

})();


demo.js
var data = [
    {date: "2015-06-12", highTemperature: 17, lowTemperature: 10, rainfall: 1.4, highWindGust:52, highWindGustBearing: 236},
    {date: "2015-06-13", highTemperature: 16, lowTemperature: 13, rainfall: 0, highWindGust: 41,highWindGustBearing: 240},
    {date: "2015-06-14", highTemperature: 16, lowTemperature: 12, rainfall: 2.4, highWindGust: 41,highWindGustBearing: 233},
    {date: "2015-06-15", highTemperature: 16, lowTemperature: 10, rainfall: 8.4, highWindGust: 40,highWindGustBearing: 261},
    {date: "2015-06-16", highTemperature: 12, lowTemperature: 4, rainfall: 1, highWindGust: 36,highWindGustBearing: 229},
    {date: "2015-06-17", highTemperature: 14, lowTemperature: 4, rainfall: 0, highWindGust: 10,highWindGustBearing: 74},
    {date: "2015-06-18", highTemperature: 15, lowTemperature: 11, rainfall: 0, highWindGust: 24,highWindGustBearing: 321},
    {date: "2015-06-19", highTemperature: 18, lowTemperature: 12, rainfall: 0, highWindGust: 35, highWindGustBearing: 340},
    {date: "2015-06-20", highTemperature: 17, lowTemperature: 15, rainfall: 9.6, highWindGust: 36, highWindGustBearing: 302},
    {date: "2015-06-21", highTemperature: 15, lowTemperature: 9, rainfall: 0.2, highWindGust: 33, highWindGustBearing: 238},
    {date: "2015-06-22", highTemperature: 14, lowTemperature: 4, rainfall: 0.4, highWindGust: 31, highWindGustBearing: 218},
    {date: "2015-06-23", highTemperature: 11, lowTemperature: 1, rainfall: 0, highWindGust: 26, highWindGustBearing: 222},
    {date: "2015-06-24", highTemperature: 11, lowTemperature: 0, rainfall: 0, highWindGust: 31, highWindGustBearing: 213},
    {date: "2015-06-25", highTemperature: 13, lowTemperature: 4, rainfall: 0, highWindGust: 43, highWindGustBearing: 241},
    {date: "2015-06-26", highTemperature: 16, lowTemperature: 9, rainfall: 0.4, highWindGust: 40, highWindGustBearing: 230},
    {date: "2015-06-27", highTemperature: 15, lowTemperature: 10, rainfall: 2.4, highWindGust: 43, highWindGustBearing: 250},
    {date: "2015-06-28", highTemperature: 16, lowTemperature: 9, rainfall: 0.6, highWindGust: 31, highWindGustBearing: 240},
    {date: "2015-06-29", highTemperature: 16, lowTemperature: 10, rainfall: 1.2, highWindGust: 45, highWindGustBearing: 250},
    {date: "2015-06-30", highTemperature: 16, lowTemperature: 11, rainfall: 1.8, highWindGust: 38, highWindGustBearing: 228},
    {date: "2015-07-01", highTemperature: 16, lowTemperature: 9, rainfall: 0.8, highWindGust: 47, highWindGustBearing: 250},
    {date: "2015-07-02", highTemperature: 14, lowTemperature: 7, rainfall: 0.2, highWindGust: 12, highWindGustBearing :47},
    {date: "2015-07-03", highTemperature: 16, lowTemperature: 7, rainfall: 1.2, highWindGust: 29, highWindGustBearing: 65},
    {date: "2015-07-04", highTemperature: 16, lowTemperature: 12, rainfall: 5.6, highWindGust: 22, highWindGustBearing: 107},
    {date: "2015-07-05", highTemperature: 15, lowTemperature: 11, rainfall: 0.2, highWindGust: 47, highWindGustBearing: 248},
    {date: "2015-07-06", highTemperature: 16, lowTemperature: 10, rainfall: 1.8, highWindGust: 28, highWindGustBearing: 312},
    {date: "2015-07-07", highTemperature: 17, lowTemperature: 11, rainfall: 41.4, highWindGust: 33, highWindGustBearing: 330},
    {date: "2015-07-08", highTemperature: 12, lowTemperature: 3, rainfall: 0.8, highWindGust: 31, highWindGustBearing: 214},
    {date: "2015-07-09", highTemperature: 10, lowTemperature: 3, rainfall: 0, highWindGust: 45, highWindGustBearing: 225},
    {date: "2015-07-10", highTemperature: 10, lowTemperature: 0, rainfall: 0, highWindGust: 28, highWindGustBearing: 215},
    {date: "2015-07-11", highTemperature: 12, lowTemperature: 0, rainfall: 0, highWindGust: 33, highWindGustBearing: 150}
];

var graph = new DailyWeatherGraph({
    data: data,
    container: document.getElementById('test'),
    width: 880,
    height: 440
});
